src/main/scala/is/hail/expr/ir/functions/CallFunctions.scala

```diff
@@ -0,0 +1,130 @@
+package is.hail.expr.ir.functions
+
+import is.hail.asm4s._
+import is.hail.expr.types._
+import is.hail.variant.Call.alleleRepr
+import is.hail.variant._
+
+object CodeAllelePair {
+  def apply(j: Code[Int], k: Code[Int]) = new CodeAllelePair(j | (k << 16))
+}
+
+class CodeAllelePair(val p: Code[Int]) {
+  val j: Code[Int] = p & 0xffff
+  val k: Code[Int] = (p >>> 16) & 0xffff
+  val nNonRefAlleles: Code[Int] =
+    j.ceq(0).mux(0, 1) + k.ceq(0).mux(0, 1)
+  val alleleIndices: (Code[Int], Code[Int]) = (j, k)
+}
+
+class CallFunctions {
+  def diploidGTIndex(j: Code[Int], k: Code[Int]): Code[Int] =
+    k * (k + 1) / 2 + j
+
+  def diploidGTIndexWithSwap(i: Code[Int], j: Code[Int]): Code[Int] = {
+    (j < i).mux(diploidGTIndex(j, i), diploidGTIndex(i, j))
+  }
+
+  def call1(aj: Code[Int], phased: Code[Boolean]): Code[Call] =
+    call(aj, phased, 1)
+
+  def call2(aj: Code[Int], ak: Code[Int], phased: Code[Boolean]): Code[Call] =
+    call(
+      phased.mux(diploidGTIndex(aj, aj + ak),
+        diploidGTIndexWithSwap(aj, ak)),
+      phased, const(2))
+
+  def call(ar: Code[Int], phased: Code[Boolean], ploidy: Code[Int]): Code[Call] = {
+    const(0) | phased.toI | ((ploidy & 3) << 1) | (ar << 3)
+  }
+
+  val throwBadPloidy = Code._fatal(s"invalid ploidy. Only support ploidy == 2")
+
+  val ploidy: IRFunction[Int] = IRFunction[Int]("ploidy", TCall(), TInt32()) {
```

Comment:
OK, two comments:

1.  Initially I wanted you to migrate the function registry over to IR rather than building a new one, but I'm increasingly convinced that this is a better way to go, primarily because ALL value representations are different in the IR.

2.  I think the register methods and friends has made quite readable function definitions in the function registry.  I'd be inclined to follow that path.  In particular, you should make it easy to register functions that just call static methods (or Scala singleton object methods) easily.